### PR TITLE
CPR Remake

### LINKED ISSUE #49: Make CPR real again.
# Description

Please, for the love of oracle, provide me with the lost love that I have been waitin to see again.

The CPR feature from nyano
When in critical condition: instead of hugging, you would perform CPR that would heal oxygen damage and stop loss of oxygen
If not done by a doctor, it would have a chance to break the patient's ribcage
# Media

### LINKED ISSUE #473: CPR Rebase and Rework Design Document
# Description
CPR was originally a Nyanotrasen feature, which allowed crewmembers to perform CPR on individuals in critical condition. When performing CPR, after a short Do-after requiring both performer and patient not move, the patient would be given the ability to breathe for a few seconds, even while in crit. Although this feature was somewhat poorly coded, and had several glaring issues with it, it did add a significant amount of depth to the game experience for players. Predating the Epi-pen and Defibrillator entirely, CPR was originally pitched as a way for non-medical crewmen to help with medical gameplay, and acted as a stopgap for the medical system being very poorly implemented. 

To state right now, I have no intention of copying the original Nyanotrasen CPR code, I do not believe it to be worth recovering. Instead I wish for it to be rewritten using existing Systems that were not originally available at the time of its creation. But after discussing the subject extensively with Old Dance Jacket, I am also putting forward a design document for a full rework of CPR, so that it is significantly more relevant to today's version of the game. My changes look to address some issues with medical gameplay that I feel are being incorrectly addressed elsewhere with destructive changes. 

# Reworked CPR Mechanics

- [ ] A Reworked CPR should add a new trait to the game, to simply be called CPR Training. All Medical Jobs automatically start with this trait for free, otherwise it must be taken in character creation. Ostensibly this would be unbalanced, were it not for PR #434. Thanks to the upcoming addition of a Trait Point system, it can be balanced as a modestly cheap positive trait.
- [ ] When a character with CPRTrainingComponent examines a crewman that is either Crit or Dying, but Not Rotting, the CPRSystem will use the Verbs system to add a CPR Verb to the right click menu.
- [ ] Clicking on the CPR Verb activates a 6-second duration, **Repeating**, **Stationary** Do-After. It also makes a popup, and plays the original Nyanotrasen CPR Sound(It's a good sound, we can reuse it even if the rest of the code is bad). 
- [ ] If the patient is rotten, cancel the do-after and make a sad popup saying something like, "{$entity} is too far gone..."
- [ ] If this Do-After is successfully completed, several things will happen

1. The Do-after will repeat itself unless interrupted
2. If the patient is in Crit, the patient will be healed of a small amount of Airloss damage
3. If the patient is "Dead But Not Rotten", it heals a small amount of Airloss damage and decrements the patient's Rot Timer by 6 seconds. This effectively means that so long as a patient is actively receiving CPR, their rot timer is paused.
4. A dead patient should metabolize a very small amount of chems anytime CPR is completed on them. 
5. If CPR can reduce a person's airloss damage such that their total damage drops below 200, there should be a very small chance that the CPR will resuscitate them back to a crit or alive state. This is rolled as probably a 1% chance each time the Do-After is completed. 

- [ ] Only ONE person should be allowed to perform CPR on a patient at a time.
- [ ] You should be prevented from performing CPR if either performer or patient have a mask, and if the patient is wearing any OuterWear item. You should have to take off a security officer's body armor before being able to access his ribcage. Obviously this by extension means CPR can't be safely performed in a vacuum.

# Okay, but they're dead?? Why reduce the rot timer??
Dead isn't a binary yes or no state, and unfortunately the game's medical system is kind of bad. It's very clear to someone with real world CPR training that the rotting system is a sort of poorly worded and portrayed abstraction of the complexity of how people die. But basically, if your heart stopped as you were reading this sentence, you wouldn't be dead. Space Station 14 would say you're dead, but clinically you aren't. If I went and started performing CPR on you while your heart was stopped, your eventual "Brain death"(the point of actual, true death) will be delayed until I run out of stamina and pass out from exhaustion. The point is to buy time for a real doctor to show up with real medical equipment, and revive you. CPR is the act of buying precious time for that doctor to arrive.

# So How Does This Improve Medical "Experience"?

1. Reworked CPR helps expand the role-play aspect of Medical. Your character gets to be the Good Samaritan by performing CPR on a coworker.
2. Gives a way for non-doctors(who buy the CPR Training trait) to positively interact with medical staff. John Greytide spent 2 of his starting trait points on CPR Training, and now he's in the medbay performing CPR on the Captain, while Urist McDoctor is frantically searching for epinephrine and bicaridine.
3. It may also provide opportunities for antagonists to potentially earn people's trust. Who in their right mind would ever suspect the guy performing CPR on the Captain?
4. Helps solve at least part of the problem of people being brought into the medbay with severe injuries, and being rejected by medical staff due to a perceived risk of rotting. This should never happen so long as a spare hand is available to perform CPR.
5. If there are zero medical supplies available at all, it gives you at least SOMETHING to do to help, rather than sit around watching people's rot timers slowly go up.
6. Gives Paramedics and Interns something extremely important to do, especially if triage is being solo'd by a Gigachad CMO with a duffel bag full of bicaridine jugs. 
7. Helps reduce reliance on Cloning/Metempsychosis. Look, not everyone likes these two machines, but they exist for a reason, and that reason is "Medical gameplay sucks". I'm not going to pretend that CPR will eliminate the need for Cloning/Metem, but by god every little thing I can do to make medical suck less, will mean people are less likely to be frustrated by the two bad-gameplay-bandaid-simulators

# But What About Morgue/StasisBed/Cryo/UnobtainiumChemistry?

A person is required to perform CPR, you can't thank a stasis bed for saving your life. This provides a substantial RP benefit to medical gameplay, encouraging teamwork, and provides a meaningful solution that doesn't involve chemists begging salvage to fulton dead space carps to the medbay. I've grown sick of other forks making the medbay jobs worse by replacing one bad solution with another, and right now the current in favor bad solution is "Unobtainium Chemistry". Utterly deranged. 

# Related PRs:

- PR #434

### PR DESCRIPTION
**Pull request by VMSolidus**
# Description

This is a re implementation and complete rewrite of the original Nyanotrasen CPR feature, this time as its own completely standalone system. Unlike the original CPR, this system requires no modification of base game code, and can be toggled on and off via CVars while the server is running.

Fixes #473
Fixes #49

# Changelog
:cl:
- add: CPR has been added to the game. People with CPR training can now perform CPR on anyone who is in either crit or dead states. 
- add: CPR Training has been added to the game as a new positive trait. All medical staff start with this trait for free.

### CODE REVIEW

File: Resources/Prototypes/Roles/Jobs/Medical/chief_medical_officer.yml [hunk 1]
```diff
@@ -44,6 +44,7 @@
     components:
     - type: PsionicBonusChance #Nyano - Summary: makes it more likely to become psionic.
       flatBonus: 0.025
+    - type: CPRTraining
```
Comment (DangerRevolution) [inline Resources/Prototypes/Roles/Jobs/Medical/chief_medical_officer.yml:47]: Why is this not;
```suggestion
    special:
    - !type:AddComponentSpecial
      components:
      - type: CPRTraining

```

Comment (DangerRevolution) [reply Resources/Prototypes/Roles/Jobs/Medical/chief_medical_officer.yml:47]: like the other roles?

Comment (DEATHB4DEFEAT) [reply Resources/Prototypes/Roles/Jobs/Medical/chief_medical_officer.yml:47]: It is

Comment (DangerRevolution) [reply Resources/Prototypes/Roles/Jobs/Medical/chief_medical_officer.yml:47]: :(

Comment (VMSolidus) [reply Resources/Prototypes/Roles/Jobs/Medical/chief_medical_officer.yml:47]: Surprise, the fact that other roles use multiple instances of AddComponentSpecial is itself incorrect and unecessary. AddComponentSpecial accepts a list of components with arguments, there is no need to add more instances of this system. I simply needed to add the CPRTraining component to the list.

**Comment by DangerRevolution** [comment]
do you plan to edit revival chance; remember you mentioning it in DS14?

**Review by DangerRevolution — APPROVED** [review]

**Comment by VMSolidus** [comment]
> do you plan to edit revival chance; remember you mentioning it in DS14?

Yea, I'm going to bump it up to like 5%.

**Review by OldDanceJacket — APPROVED** [review]

**Pull request merged**